iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -596,6 +596,97 @@ static LogicalResult setRootConfig(
 }
 
+static LogicalResult setX86DefaultParallelReductionTileSizes(
+    linalg::GenericOp genericOp, unsigned numLoops,
+    ArrayRef<int64_t> flowTileSizes, ArrayRef<int64_t> minTileSizes,
+    ArrayRef<int64_t> maxTileSizes, SmallVectorImpl<int64_t> &parallelTileSizes,
+    SmallVectorImpl<int64_t> &reductionTileSizes) {
+  parallelTileSizes.append(numLoops, 0);
+  Optional<SmallVector<int64_t, 4>> staticLoopRanges =
+      genericOp.getStaticLoopRanges();
+  for (auto loopNum : llvm::seq<unsigned>(0, numLoops)) {
+    if (flowTileSizes[loopNum]) {
+      parallelTileSizes[loopNum] =
+          getMaxTileSize(0, flowTileSizes[loopNum], minTileSizes[loopNum],
+                         minTileSizes[loopNum]);
+    } else {
+      // If the flow level tile size is zero, and static loop range is 0 as
+      // well, set the tile sizes here to zero as well.
+      parallelTileSizes[loopNum] =
+          (staticLoopRanges && staticLoopRanges.getValue()[loopNum] == 1)
+              ? 0
+              : minTileSizes[loopNum];
+    }
+  }
+
+  splitParallelAndReductionTiles(genericOp, parallelTileSizes,
+                                 reductionTileSizes);
+  return success();
+}
+
+/// Returns true if the operation is a GenericOp implementing a supported
+/// transposition.
+static bool isSupportedTransposeOp(linalg::GenericOp genericOp) {
+  // Check that the op has at least 2 dimensions.
+  if (genericOp.getNumLoops() < 2) return false;
+
+  // Check that the op has only one input and one output.
+  // TODO(diegocaballero): Generalize to multiple inputs.
+  if ((genericOp.getNumInputs() != 1) || (genericOp.getNumOutputs() != 1))
+    return false;
+
+  // Check that all the iterators are parallel.
+  auto iterator_types = genericOp.iterator_types();
+  if (std::any_of(
+          iterator_types.begin(), iterator_types.end(),
+          [](Attribute attr) { return !mlir::isParallelIterator(attr); }))
+    return false;
+
+  // Check that the two indexing maps are a permutation of each other.
+  auto indexing_maps = genericOp.getIndexingMaps();
+  if (indexing_maps.size() != 2) return false;
+  return (indexing_maps[0].isIdentity() && indexing_maps[1].isPermutation()) ||
+         (indexing_maps[0].isPermutation() && indexing_maps[1].isIdentity());
+}
+
+// Sets specific parallel and reduction tile sizes for tranpose operations.
+// Returns failure if 'genericOp' does not implement a supported transpose
+// operation or if there are no target features that would benefit from this
+// specific tiling (e.g., '+avx2').
+static LogicalResult setX86TransposeParallelReductionTileSizes(
+    linalg::GenericOp genericOp, unsigned numLoops,
+    ArrayRef<int64_t> flowTileSizes, ArrayRef<int64_t> minTileSizes,
+    ArrayRef<int64_t> maxTileSizes, SmallVectorImpl<int64_t> &parallelTileSizes,
+    SmallVectorImpl<int64_t> &reductionTileSizes) {
+  if (!hasAVX2Features(genericOp) || !isSupportedTransposeOp(genericOp))
```

Comment:
Nit: So the `failure()` here is reserved for actual failures which need to propagated and caught. For failure to set the configuration, its better to just return `success()`. All other methods are supposed to check that the configuration is not set before proceeding.

Reply:
Thanks for letting me know! In general, I've seen `LogicalResult` being used in a more generic way in LLVM/MLIR. I'm catching the failure in line 722 and doing something else. It seems a bit weird to return success when the method is failing to do what it's supposed to. Would it make sense to return a `bool` instead? I don't see any of these methods failing in a try/catch "exception-raising" mode, if I understand that correctly from your message. WDYT?

Otherwise, how could I check that the configuration is not set in line 722? Checking that the parallel and reduction tiles are not set?

Reply:
`bool` doesnt work for the case where it is an actual failure. `LogicalResult` is to catch that. That the method didnt set a configuration is part of the contract of such methods, i.e. it either sets a configuration or not, is expected behavior and is not a failure of the method. Its probably matter of definition of failure mode.

To check if a configuration is set, it is expected to do something like this https://github.com/google/iree/blob/main/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L744 . A method tries to set the configuration if the configuration is not set.

This avoids having to track a bunch of `if-elseif-elseif-elsif....` kind of logic which could get really complicated here. If you miss the chain at any point it would be hard to find where. So instead the setup is to do a 
```
set...Config(...)
set...Config(...)
...
set...Config(...)
```

with only one of those actually setting the configuration and everything else just returning early if the configuration is already set.